src/contracts/pods/EigenPod.sol

```diff
@@ -485,256 +505,126 @@ contract EigenPod is IEigenPod, Initializable, ReentrancyGuardUpgradeable, Eigen
+        // - validator submitted twice during this checkpoint
+        // - validator withdrawal credentials verified after checkpoint starts, then submitted
+        //   as a checkpoint proof
+        // TODO - we might want to "skip" and emit an event, rather than revert?
         require(
-            validatorInfo.status == VALIDATOR_STATUS.ACTIVE, 
-            "EigenPod.verifyBalanceUpdate: Validator not active"
+            validatorInfo.mostRecentBalanceUpdateTimestamp < beaconTimestamp,
+            "EigenPod._verifyCheckpointProof: validator already proven for this checkpoint"
         );
-
-        // 3. Balance updates should only be made before a validator is fully withdrawn. 
-        // -- A withdrawable validator may not have withdrawn yet, so we require their balance is nonzero
-        // -- A fully withdrawn validator should withdraw via verifyAndProcessWithdrawals
-        if (validatorFields.getWithdrawableEpoch() <= _timestampToEpoch(oracleTimestamp)) {
-            require(
-                validatorEffectiveBalanceGwei > 0,
-                "EigenPod.verifyBalanceUpdate: validator is withdrawable but has not withdrawn"
-            );
-        }
-
-        // Verify passed-in validatorFields against verified beaconStateRoot:
-        BeaconChainProofs.verifyValidatorFields({
+        
+        // Verify validator balance against beaconStateRoot
+        uint64 prevBalanceGwei = validatorInfo.restakedBalanceGwei;
+        uint64 newBalanceGwei = BeaconChainProofs.verifyValidatorBalance({
             beaconStateRoot: beaconStateRoot,
-            validatorFields: validatorFields,
-            validatorFieldsProof: validatorFieldsProof,
-            validatorIndex: validatorIndex
+            validatorIndex: uint40(validatorInfo.validatorIndex),
+            proof: proof
         });
-
-        // Done with proofs! Now update the validator's balance and send to the EigenPodManager if needed
-
-        uint64 currentRestakedBalanceGwei = validatorInfo.restakedBalanceGwei;
-        uint64 newRestakedBalanceGwei;
-        if (validatorEffectiveBalanceGwei > MAX_RESTAKED_BALANCE_GWEI_PER_VALIDATOR) {
-            newRestakedBalanceGwei = MAX_RESTAKED_BALANCE_GWEI_PER_VALIDATOR;
-        } else {
-            newRestakedBalanceGwei = validatorEffectiveBalanceGwei;
-        }
         
-        // Update validator balance and timestamp, and save to state:
-        validatorInfo.restakedBalanceGwei = newRestakedBalanceGwei;
-        validatorInfo.mostRecentBalanceUpdateTimestamp = oracleTimestamp;
-        _validatorPubkeyHashToInfo[validatorPubkeyHash] = validatorInfo;
+        // Update validator state. If their new balance is 0, they are marked `WITHDRAWN`
+        validatorInfo.restakedBalanceGwei = newBalanceGwei;
+        validatorInfo.mostRecentBalanceUpdateTimestamp = beaconTimestamp;
+        if (newBalanceGwei == 0) {
+            activeValidatorCount--;
+            validatorInfo.status = VALIDATOR_STATUS.WITHDRAWN;
```

Comment:
probably want to emit an event when this happens?

Reply:
yespls!

Reply:
[7aa93fb](https://github.com/Layr-Labs/eigenlayer-contracts/pull/515/commits/7aa93fb4216af788d48eed2cbc81bfbcc7b42d6e)